### PR TITLE
s4701574 - Stable Diffusion for Image Generation with ADNI data

### PR DESCRIPTION
**Pull request by randy622**
Hello, 

My name is Randy Abou Chakra, and my student ID is 47015746. This is a pull request to merge my topic-recognition into Shake's topic-recognition. The task I picked was Stable Diffusion to generate brain images using the ADNI dataset. My results are semi-successful; the generated images have the texture of a brain and the general shape of one, but they fail to capture the specific details of a brain like the ones that could be seen in the ADNI data. In addition, I tested the Stable Diffusion model with a simple autoencoder and then with a VQ-VAE, and the image generation drastically improved. The files and folders included in this pull request are:

StableDiffusion-47015746: folder of the project

dataset.py: includes dataloader method
modules.py: includes model architecture for the Stable Diffusion model, which has the VQ-VAE and U-Net
train.py: includes the methods to train the models (specifically, the VQ-VAE and the denoising U-Net)
predict.py: includes image generation method given a trained Stable Diffusion model. 
driver.py: includes calls to the methods necessary to train Stable Diffusion model and then generate images
resources: includes images and gifs used in the README

Thank you,
Randy

### CODE REVIEW

**Comment by gayanku** [comment]
*****************************************************************
**This is an initial inspection, no action is required at this point**
Stable Diffusion ADNI brain data ---- Hard Difficulty
| Category	| | Marks | Comments |	
| :---  | :---: | :---: | :---  |		
| Algorithm solves the problem 		| 5 | 2 | Unconditional, but this is fine. Uses a VQ-VAE.  DiffusionModel has some issues	|																					
| Implementation functions as intended	| 3 | 1.5 |  Images not too clear. |
| Good design 					| 1 | 1 | Modular, Reusable |
| Commenting 					| 1 | 1 | Meaningful docstrings and comments |
| Algorithm above Normal Difficulty 	| 5 | 5 | |
| Algorithm is Hard difficulty 			| 5 | 5 | Hard Difficulty |
| Section IV : Max mark 20 from 20		|   | **15.5**   | |

**TSNE / UMAP:** 
Not provided.
While no marks have been reduced, the task is to train one model on the whole dataset, and by plotting the embedding of style code, it is expected to see somehow clustered patterns for different classes.

**Suggestions:**
Your DiffusionModel should use the VQ/VAE to encode to latent space, then to the removal in latent space (of the selected time steps). Finally, at the end of the time steps, decode back to the image.  Your forward() is using VAE.encode() in each time step. There is also no call to the VAE.decode() anywhere in your train.py

**Comment by YashMitttal** [comment]
Some additional comments:

- Clarify Model Evaluation Metric: Since the final generated images do not fully resemble the training data, consider defining a metric or quantitative measure to evaluate the resemblance (e.g., Fréchet Inception Distance or another similarity metric).
- Clearly, as Gayan said - the progressive diffusion still needs some work
- Also, the image quality is not the best, and you should have explained why you chose certain hyperparameters and data preprocessing metrics.

Overall, the project is well-structured, with a clear overview and architecture explanation.

**Comment by shakes76** [comment]
Good work!
